src/pyhf/tensor/tensorflow_backend.py

```diff
@@ -195,26 +186,19 @@ def reshape(self, tensor, newshape):
         return tf.reshape(tensor, newshape)
 
     def divide(self, tensor_in_1, tensor_in_2):
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
         return tf.divide(tensor_in_1, tensor_in_2)
 
     def log(self, tensor_in):
-        tensor_in = self.astensor(tensor_in)
         return tf.math.log(tensor_in)
 
     def exp(self, tensor_in):
-        tensor_in = self.astensor(tensor_in)
         return tf.exp(tensor_in)
 
     def stack(self, sequence, axis=0):
         return tf.stack(sequence, axis=axis)
 
     def where(self, mask, tensor_in_1, tensor_in_2):
-        mask = self.astensor(mask)
-        tensor_in_1 = self.astensor(tensor_in_1)
-        tensor_in_2 = self.astensor(tensor_in_2)
-        return mask * tensor_in_1 + (1 - mask) * tensor_in_2
+        return tf.compat.v2.where(tf.cast(mask, tf.bool), tensor_in_1, tensor_in_2)
```

Comment:
@kratsg Your comment got removed when I combined two lines, so answering here again.

> why is there a cast here? The mask should be `bool` already.

I think that it is better to just cast it to make sure. There are times like the following where the mask seems like a valid `Bool`

https://github.com/diana-hep/pyhf/blob/c8a94f4d5a11edfdd5a28184153e1ef4b076e638/tests/test_tensor.py#L48-L50

but TensorFlow doesn't see it as such if the `tf.cast(mask, tf.bool)` is removed. So I think that it should just be used to ensure the right Tensor `dtype`.

Reply:
https://github.com/diana-hep/pyhf/blob/d9aeac8bd3b7bd384342ae2e0e7547b78714e66c/src/pyhf/tensor/tensorflow_backend.py#L124

That's because the default cast is to float. You should use `tb.astensor(tensor, dtype='bool')` for the mask input.

tl;dr - the test is wrong.

Reply:
```
>>> import pyhf
>>> pyhf.set_backend(pyhf.tensor.tensorflow_backend())
>>> from pyhf import tensorlib as tb
>>> tb.astensor([1,0,1])
array([1., 0., 1.])
>>> tb.astensor([1,0,1], dtype='float')
array([1., 0., 1.])
>>> tb.astensor([1,0,1], dtype='bool')
array([ True, False,  True])
```

Reply:
Not sure where my last comment went (somewhere into the void?) but yes, it is clear as to the _why_ of the behavior by not to the reasonableness. I have changed the test, but I think it is better to keep the cast there as a user could quite legitimately view the `tb.astensor([1,0,1])` as a valid form of Bool (as NumPy does). The cast should be faster then the `self.astensor` call that we have previously been doing, so it seems like keeping this cast in there should be reasonable and keep things safe.

Reply:
> as a valid form of Bool (as NumPy does)

It's not that numpy handles it as a valid form, it type-casts for you implicitly as part of the masking - but that might not be good behavior. We should have users be explicit about the type of tensor they're dealing with (bools, floats, ints) - because they all store memory in different ways and are allocated differently.

> The cast should be faster then the `self.astensor` call that we have previously been doing

Then we should change the tensorflow `astensor` to use the `cast` then... if it's faster.

Reply:
> Then we should change the tensorflow `astensor` to use the `cast` then... if it's faster.

We do already

https://github.com/diana-hep/pyhf/blob/f952b8bca853e09259351b2ebc2a09c918c23e6f/src/pyhf/tensor/tensorflow_backend.py#L152-L153

doing this just skips the rest of the logic of `astensor`. I will still work on changing the calls in other parts, as needed, to explicitly request Bools.